File: MPI/02.PingPong/pingpong-bench.py
from mpi4py import MPI

number_of_messages = 50
buffer = 0.0
status = MPI.Status()
comm_world = MPI.COMM_WORLD
my_rank = comm_world.Get_rank()
num_procs = comm_world.Get_size()

if(num_procs == 2):
  start = MPI.Wtime();
  for i in range(number_of_messages):
    if (my_rank == 0):
      comm_world.send(buffer, dest=1, tag=17)
      buffer = comm_world.recv(source=1, tag=23, status=status)
    elif (my_rank == 1):
      buffer = comm_world.recv(source=0, tag=17, status=status)
      comm_world.send(buffer, dest=0, tag=23)
  finish = MPI.Wtime();

  if (my_rank == 0):
    msg_transfer_time = ((finish - start) / (2 * number_of_messages)) * 1e6
    print(f"Time for one message: {msg_transfer_time:f} micro seconds.")
else:
  if(my_rank == 0):
    print(f"Number of processor is not 2, aborting.")

File: MPI/02.PingPong/pingpong-bench.cpp
#include <mpi.h>

int main(int argc, char* argv[]){
  int my_rank, num_procs, buffer;
  const int number_of_messages = 50;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
  MPI_Comm_size(MPI_COMM_WORLD, &num_procs);

  if(num_procs == 2){
    double start = MPI_Wtime();
    for(int i = 0; i < number_of_messages; i++){
      if(my_rank == 0){
        MPI_Send(&buffer, 1, MPI_INT, 1, 17, MPI_COMM_WORLD);
        MPI_Recv(&buffer, 1, MPI_INT, 1, 23, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
      } else {
        MPI_Recv(&buffer, 1, MPI_INT, 0, 17, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        MPI_Send(&buffer, 1, MPI_INT, 0, 23, MPI_COMM_WORLD);
      }
    }
    double finish = MPI_Wtime();

    if(my_rank == 0){
      double msg_transfer_time = ((finish - start) / (2 * number_of_messages)) * 1e6;
      printf("Time for one message: %f micro seconds.\n", msg_transfer_time);
    }
  } else {
    if(my_rank == 0){
      printf("Number of processor is not 2, aborting.\n");
    }
  }

  MPI_Finalize();
}